Lobby player list in alphabetical order

Ninetailed opened the pull request:
Changed the lobby player list to alphabetical order, rather than the unordered mess it was before. There's theoretically a performance hit for doing this but it wasn't enough that I noticed during testing with two users. I don't have a good way of testing with more, but it only happens before the round starts.

![alphabetical-lobby-list](https://user-images.githubusercontent.com/1348714/98827576-f015fb00-242e-11eb-9241-d17391deeed1.png)

🆑 Ninetailed
rscadd: Lobby player list in alphabetical order
/🆑

EDIT: New impementation
Sorts GLOB.player_list by client key

This is done by using ADD_SORTED to add newly logged in players to the global list. This results in a single O(n) performance hit on login. This is an infrequent enough event that this should be acceptable, but may manifest as a transient lag spike with large numbers of players.

This will result in the Lobby player list appearing in alphabetical order, and may also have the same effect on certain admin tools such as the debug player list.

As this is done on login, if a mob is transferred to a different client key, the list will not be updated. As this change primarily targets the pre-game lobby, before any such changes will be made, I consider this case to be out of scope.

Ninetailed commented:
Updated to use the ADD_SORTED macro in response to feedback.

Spookerton commented:
Hmm. I feel like making player_list sorted would be preferable since that means doing a sorted add only once per login.

Ninetailed commented:
Probably doable. I'll look into it.

Ninetailed committed:
Sorts GLOB.player_list by client key

This is done by using ADD_SORTED to add newly logged in players to the
global list. This results in a single O(n) performance hit on login.
This is an infrequent enough event that this should be acceptable, but
may manifest as a transient lag spike with large numbers of players.

This will result in the Lobby player list appearing in alphabetical
order, and may also have the same effect on certain admin tools such as
the debug player list.

As this is done on login, if a mob is transferred to a different client
key, the list will not be updated. As this change primarily targets the
pre-game lobby, before any such changes will be made, I consider this
case to be out of scope.

Apply suggestions from code review

Co-authored-by: Spookerton <spkrtn@pm.me>, Mucker <colehoul@gmail.com>

Ninetailed commented:
Changed to pre-sorting player_list. Original post updated to reflect.

MuckerMayhem commented on code/_helpers/cmp.dm:
```diff
@@ -84,3 +84,13 @@
 /proc/cmp_planelayer(atom/A, atom/B)
 	return (B.plane - A.plane) || (B.layer - A.layer)
+
+/proc/cmp_mob_key(var/mob/A, var/mob/B)
+	if (isnull(A))
+		if (isnull(B))
+			return 0
+		else
+			return -1
+	else if (isnull(B))
+		return 1
+	return cmp_text_asc(A.key, B.key)
```
```
/proc/cmp_mob_key(var/mob/A, var/mob/B)
	if (!A && !B)
		return 0

	if (!A && B)
		return -1

	if (A && !B)
		return 1

	return cmp_text_asc(A.key, B.key)
```
Is probably an easier to read implementation, if I understood what you're doing here correctly.

Spookerton replied:
```suggestion
/proc/cmp_mob_key(mob/A, mob/B)
	if (!A)
		if (!B)
			return 0
		return -1
	if (!B)
		return 1
	return sorttext(B.key, A.key)
```

Spookerton commented on code/modules/mob/login.dm:
```diff
@@ -58,7 +58,10 @@
 
-	GLOB.player_list |= src
+	// Add to player list if missing
+	if (GLOB.player_list.Find(src) == 0)
```
```suggestion
	if (!GLOB.player_list.Find(src))
```

Spookerton commented on code/modules/mob/new_player/login.dm:
```diff
@@ -13,7 +13,10 @@
-	GLOB.player_list |= src
+
+	// Add to player list if missing
+	if (GLOB.player_list.Find(src) == 0)
```
```suggestion
	if (!GLOB.player_list.Find(src))
```

MuckerMayhem commented on code/_helpers/cmp.dm:
```diff
@@ -84,3 +84,13 @@
 
 /proc/cmp_planelayer(atom/A, atom/B)
 	return (B.plane - A.plane) || (B.layer - A.layer)
+
+/proc/cmp_mob_key(var/mob/A, var/mob/B)
+	if (isnull(A))
+		if (isnull(B))
+			return 0
+		else
+			return -1
+	else if (isnull(B))
+		return 1
+	return cmp_text_asc(A.key, B.key)
```
```
/proc/cmp_mob_key(var/mob/A, var/mob/B)
	if (!A && !B)
		return 0

	if (!A && B)
		return -1

	if (A && !B)
		return 1

	return cmp_text_asc(A.key, B.key)
```
Is probably an easier to read implementation, if I understood what you're doing here correctly.

MuckerMayhem replied:
Spook disagrees on this one, so we will default to his suggestion, for he is wise.

Spookerton commented:
Squash pleez

Ninetailed commented:
I have squished it! It is flat now.
